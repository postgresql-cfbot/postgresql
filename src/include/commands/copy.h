/*-------------------------------------------------------------------------
 *
 * copy.h
 *	  Definitions for using the POSTGRES copy command.
 *
 *
 * Portions Copyright (c) 1996-2024, PostgreSQL Global Development Group
 * Portions Copyright (c) 1994, Regents of the University of California
 *
 * src/include/commands/copy.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef COPY_H
#define COPY_H

#include "commands/copyapi.h"
#include "nodes/parsenodes.h"
#include "parser/parse_node.h"
#include "tcop/dest.h"

extern void DoCopy(ParseState *pstate, const CopyStmt *stmt,
				   int stmt_location, int stmt_len,
				   uint64 *processed);

extern void ProcessCopyOptions(ParseState *pstate, CopyFormatOptions *opts_out, bool is_from, List *options);
extern CopyFromState BeginCopyFrom(ParseState *pstate, Relation rel, Node *whereClause,
								   const char *filename,
								   bool is_program, copy_data_source_cb data_source_cb, List *attnamelist, List *options);
extern void EndCopyFrom(CopyFromState cstate);
extern bool NextCopyFrom(CopyFromState cstate, ExprContext *econtext,
						 Datum *values, bool *nulls);
extern void CopyFromErrorCallback(void *arg);
extern char *CopyLimitPrintoutLength(const char *str);

extern uint64 CopyFrom(CopyFromState cstate);

extern DestReceiver *CreateCopyDestReceiver(void);

/*
 * internal prototypes
 */
extern CopyToState BeginCopyTo(ParseState *pstate, Relation rel, RawStmt *raw_query,
							   Oid queryRelId, const char *filename, bool is_program,
							   copy_data_dest_cb data_dest_cb, List *attnamelist, List *options);
extern void EndCopyTo(CopyToState cstate);
extern uint64 DoCopyTo(CopyToState cstate);
extern List *CopyGetAttnums(TupleDesc tupDesc, Relation rel,
							List *attnamelist);

#endif							/* COPY_H */
