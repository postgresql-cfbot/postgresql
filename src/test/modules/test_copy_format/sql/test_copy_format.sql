CREATE EXTENSION test_copy_format;
CREATE TABLE public.test (a INT, b INT, c INT);
INSERT INTO public.test VALUES (1, 2, 3), (12, 34, 56), (123, 456, 789);
COPY public.test FROM stdin WITH (
	option_before 'before',
	format 'test_copy_format',
	option_after 'after'
);
\.
COPY public.test TO stdout WITH (
	option_before 'before',
	format 'test_copy_format',
	option_after 'after'
);
