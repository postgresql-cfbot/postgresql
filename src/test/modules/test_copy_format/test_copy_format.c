/*--------------------------------------------------------------------------
 *
 * test_copy_format.c
 *		Code for testing custom COPY format.
 *
 * Portions Copyright (c) 2024, PostgreSQL Global Development Group
 *
 * IDENTIFICATION
 *		src/test/modules/test_copy_format/test_copy_format.c
 *
 * -------------------------------------------------------------------------
 */

#include "postgres.h"

#include "commands/copy.h"
#include "commands/defrem.h"

PG_MODULE_MAGIC;

static bool
CopyFromProcessOption(CopyFromState cstate, DefElem *defel)
{
	ereport(NOTICE,
			(errmsg("CopyFromProcessOption: \"%s\"=\"%s\"",
					defel->defname, defGetString(defel))));
	return true;
}

static int16
CopyFromGetFormat(CopyFromState cstate)
{
	ereport(NOTICE, (errmsg("CopyFromGetFormat")));
	return 0;
}

static void
CopyFromStart(CopyFromState cstate, TupleDesc tupDesc)
{
	ereport(NOTICE, (errmsg("CopyFromStart: natts=%d", tupDesc->natts)));
}

static bool
CopyFromOneRow(CopyFromState cstate, ExprContext *econtext, Datum *values, bool *nulls)
{
	ereport(NOTICE, (errmsg("CopyFromOneRow")));
	return false;
}

static void
CopyFromEnd(CopyFromState cstate)
{
	ereport(NOTICE, (errmsg("CopyFromEnd")));
}

static const CopyFromRoutine CopyFromRoutineTestCopyFormat = {
	.type = T_CopyFromRoutine,
	.CopyFromProcessOption = CopyFromProcessOption,
	.CopyFromGetFormat = CopyFromGetFormat,
	.CopyFromStart = CopyFromStart,
	.CopyFromOneRow = CopyFromOneRow,
	.CopyFromEnd = CopyFromEnd,
};

static bool
CopyToProcessOption(CopyToState cstate, DefElem *defel)
{
	ereport(NOTICE,
			(errmsg("CopyToProcessOption: \"%s\"=\"%s\"",
					defel->defname, defGetString(defel))));
	return true;
}

static void
CopyToSendCopyBegin(CopyToState cstate)
{
	ereport(NOTICE, (errmsg("CopyToGetFormat")));
}

static void
CopyToStart(CopyToState cstate, TupleDesc tupDesc)
{
	ereport(NOTICE, (errmsg("CopyToStart: natts=%d", tupDesc->natts)));
}

static void
CopyToOneRow(CopyToState cstate, TupleTableSlot *slot)
{
	ereport(NOTICE, (errmsg("CopyToOneRow: tts_nvalid=%u", slot->tts_nvalid)));
}

static void
CopyToEnd(CopyToState cstate)
{
	ereport(NOTICE, (errmsg("CopyToEnd")));
}

static const CopyToRoutine CopyToRoutineTestCopyFormat = {
	.type = T_CopyToRoutine,
	.CopyToProcessOption = CopyToProcessOption,
	.CopyToSendCopyBegin = CopyToSendCopyBegin,
	.CopyToStart = CopyToStart,
	.CopyToOneRow = CopyToOneRow,
	.CopyToEnd = CopyToEnd,
};

PG_FUNCTION_INFO_V1(test_copy_format);
Datum
test_copy_format(PG_FUNCTION_ARGS)
{
	bool		is_from = PG_GETARG_BOOL(0);

	ereport(NOTICE,
			(errmsg("test_copy_format: is_from=%s", is_from ? "true" : "false")));

	if (is_from)
		PG_RETURN_POINTER(&CopyFromRoutineTestCopyFormat);
	else
		PG_RETURN_POINTER(&CopyToRoutineTestCopyFormat);
}
