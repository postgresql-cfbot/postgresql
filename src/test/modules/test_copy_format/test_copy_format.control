comment = 'Test code for custom COPY format'
default_version = '1.0'
module_pathname = '$libdir/test_copy_format'
relocatable = true
