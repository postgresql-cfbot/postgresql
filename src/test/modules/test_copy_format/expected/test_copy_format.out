CREATE EXTENSION test_copy_format;
CREATE TABLE public.test (a INT, b INT, c INT);
INSERT INTO public.test VALUES (1, 2, 3), (12, 34, 56), (123, 456, 789);
COPY public.test FROM stdin WITH (
	option_before 'before',
	format 'test_copy_format',
	option_after 'after'
);
NOTICE:  test_copy_format: is_from=true
NOTICE:  CopyFromProcessOption: "option_before"="before"
NOTICE:  CopyFromProcessOption: "option_after"="after"
NOTICE:  CopyFromGetFormat
NOTICE:  CopyFromStart: natts=3
NOTICE:  CopyFromOneRow
NOTICE:  CopyFromEnd
COPY public.test TO stdout WITH (
	option_before 'before',
	format 'test_copy_format',
	option_after 'after'
);
NOTICE:  test_copy_format: is_from=false
NOTICE:  CopyToProcessOption: "option_before"="before"
NOTICE:  CopyToProcessOption: "option_after"="after"
NOTICE:  CopyToGetFormat
NOTICE:  CopyToStart: natts=3
NOTICE:  CopyToOneRow: tts_nvalid=3
NOTICE:  CopyToOneRow: tts_nvalid=3
NOTICE:  CopyToOneRow: tts_nvalid=3
NOTICE:  CopyToEnd
